### PR TITLE
(books.json,index.html,index.js,books-data.js,books-search.js):Develo…

### PR DESCRIPTION
**Pull request by jsequea**
Development  of card WBooks - Primer webcomponent

https://trello.com/c/BZGrYVKQ/19-wbooks-primer-webcomponent

image

### CODE REVIEW

File: components/books-data.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import { LitElement, html } from "lit-element";
+import styles from "../styles";
```
Comment (dsandoval11) [inline components/books-data.js]: defines a style file for each component

File: components/books-data.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import { BookSearch } from "./books-search";
+
+export class BookData extends LitElement {
```
Comment (dsandoval11) [inline components/books-data.js]: Don't export the class

File: components/books-data.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  static get properties() {
+    return {
+      data: [],
+      books: [],
+      allBooks: { type: Array },
+      allRestore: { type: Array }
```
Comment (dsandoval11) [inline components/books-data.js]: These properties aren't received in the component, also you don't use the `books` property
```suggestion
      data: { attribute: false },
      allBooks: { attribute: false },
      allRestore: { attribute: false }
```

File: components/books-data.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+  constructor() {
+    super();
+    this.allBooks = [];
+    this.allRestore=[];
+    this.cargaLibros();
+    this.books = [];
+    this.addEventListener("ApiData", (e) => {
+      this._dataFormat(e.detail.data);
+    });
```
Comment (dsandoval11) [inline components/books-data.js]: don't define listeners directly in the constructor, you can do this in the `firstUpdate` method

File: components/books-data.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      allRestore: { type: Array }
+    };
+  }
+
+  constructor() {
+    super();
+    this.allBooks = [];
+    this.allRestore=[];
+    this.cargaLibros();
+    this.books = [];
```
Comment (dsandoval11) [inline components/books-data.js]: first set the values and then call the functions
```suggestion
    this.allBooks = [];
    this.allRestore=[];
    this.books = [];
    this.cargaLibros();
```

File: components/books-data.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    this.dataFromSearch();
+  }
+
+  static get styles() {
+    return [styles];
+  }
+
+  async cargaLibros() {
+    fetch("../books.json").then((respuesta) => respuesta.json());
+  }
```
Comment (dsandoval11) [inline components/books-data.js]: this is not necessary

File: components/books-data.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      })
+    );
+  }
+
+  getData() {
+    fetch("../books.json")
+      .then((response) => {
+        if (response.ok) return response.json();
+        return Promise.reject(response);
+      })
```
Comment (dsandoval11) [inline components/books-data.js]: if response has an error, it should throw an exception, you can omit this part and just return the json
```suggestion
      .then((response) => response.json())
```

File: components/books-data.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  _dataFormat(data) {
+    let books = [];
+    data["results"].forEach((book) => {
+      books.push({
+        img: book.img,
+        book_title: book.book_title,
+        author: book.author,
+      });
+    });
+    this.allBooks = books;
```
Comment (dsandoval11) [inline components/books-data.js]: use the `map` function to avoid defining a temp variable

File: components/books-data.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  dataFromSearch(){
+    this.addEventListener('datasave', (e)=> this.processData(e.detail.data));
+  }
+
+  processData(data){
+    if(data.length==0){
+      this.allRestore=this.allBooks;
+    }else{
+      this.allRestore=data;
+    }
```
Comment (dsandoval11) [inline components/books-data.js]: You can use a ternary operator

File: components/books-data.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+                <img class="cover" src="${book.img}" />
+              </figure>
+              <div class="contenido">
+                <div class="encabezado">
+                  <h1 class="titulo">${book.book_title}</h1>
+                  <p class="caracteristicas">${book.author}</p>
+                </div>
+              </div>
+            </div>
+
```
Comment (dsandoval11) [inline components/books-data.js]: Fix the indentation 
```suggestion
          <div class="flexcontainer">
            <figure class="portada">
              <img class="cover" src="${book.img}" />
            </figure>
            <div class="contenido">
              <div class="encabezado">
                <h1 class="titulo">${book.book_title}</h1>
                <p class="caracteristicas">${book.author}</p>
              </div>
            </div>
          </div>
```

File: components/books-search.js [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+export class BookSearch extends LitElement {
+
+    static get properties(){
+        return{
+            data: { type: Array },
+            result: { type:Array},
+            dataFromBook: { type:Array } 
+
+        }
+    }
```
Comment (dsandoval11) [inline components/books-search.js]: Always use the same indentation (2 spaces). Fix it in the whole file

File: components/books-search.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+    filterData(){
+        const imgNotFound="https://goo.su/crvbz";
+        const input=this.shadowRoot.querySelector("#search").value.toLowerCase();
+        this.result=[];
+        this.dataFromBook.map(book => {
+            const name=book.book_title.toLowerCase();
+            if(name.indexOf(input) !== -1)
+            this.result=[...this.result, book]
+        });
```
Comment (dsandoval11) [inline components/books-search.js]: use the `filter` function

File: components/books-search.js [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    render(){
+        return html`
+        <div class="contenedor">
+        <input  @keyup=${this.filterData} type="text" id="search" placeholder="Buscá por título del libro..."/>
+        <svg xmlns="http://www.w3.org/2000/svg" class="input-icon" viewBox="0 0 20 20" fill="currentColor">
+		<path fill-rule="evenodd" d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z" clip-rule="evenodd" />
+	</svg>
+            
+
+        </div>
```
Comment (dsandoval11) [inline components/books-search.js]: Fix the indentation

File: package.json [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  "main": "index.js",
+  "scripts": {
+    "start": "web-dev-server --node-resolve --open --watch"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "devDependencies": {
+    "@web/dev-server": "^0.1.34"
+  }
```
Comment (dsandoval11) [inline package.json]: Adds the `lit-element` dependency

File: styles.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+  }
+  
+  main {
+    width: 70%;
+    display: flex;
+    flex-wrap: wrap;
+    place-content: space-around flex-end;
+    justify-content: center;
+    margin-left: 150px;
+  }
```
Comment (dsandoval11) [inline styles.js]: if possible, always use classes

File: components/books-data.js (hunk 1 above)
Comment (jsequea) [reply components/books-data.js]: Solved

File: components/books-data.js [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+    this.allRestore=this.allBooks;
+    console.log(this.allBooks);
+  }
+
+  dataFromSearch(){
+    this.addEventListener('datasave', (e)=> this.processData(e.detail.data));
+  }
+
+  processData(data){
+    data.length==0 ? this.allRestore=this.allBooks:this.allRestore=data;
```
Comment (dsandoval11) [inline components/books-data.js]:
```suggestion
    this.allRestore = data.length == 0 ? this.allBooks : data;
```

File: components/books-data.js [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+                <img class="cover" src="${book.img}" />
+              </figure>
+              <div class="contenido">
+                <div class="encabezado">
+                  <h1 class="titulo">${book.book_title}</h1>
+                  <p class="caracteristicas">${book.author}</p>
+                </div>
+              </div>
+            </div>
+
```
Comment (dsandoval11) [inline components/books-data.js]: fix the indentation

File: components/books-data.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+    <div class="main">
+      ${this.allRestore.map((book) => html`
+            <div class="flexcontainer">
+              <figure class="portada">
+                <img class="cover" src="${book.img}" />
+              </figure>
+              <div class="contenido">
+                <div class="encabezado">
+                  <h1 class="titulo">${book.book_title}</h1>
+                  <p class="caracteristicas">${book.author}</p>
```
Comment (dsandoval11) [inline components/books-data.js]: I recommend that you name the css classes in English

File: components/books-search.js [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+export class BookSearch extends LitElement {
+
+    static get properties(){
+        return{
+            data: { type: Array },
+            result: { type:Array},
+            dataFromBook: { type:Array } 
+        }
+    }
```
Comment (dsandoval11) [inline components/books-search.js]: always use the same indention, preferably 2 spaces

File: components/books-search.js [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+        this.dispatchEvent(new CustomEvent('datasave', {detail:{data}, bubbles:true, composed:true }));
+    }
+
+    render(){
+        return html`
+        <div class="contenedor">
+            <input  @keyup=${this.filterData} type="text" id="search" placeholder="Buscá por título del libro..."/>
+            <svg xmlns="http://www.w3.org/2000/svg" class="input-icon" viewBox="0 0 20 20" fill="currentColor">
+                <path fill-rule="evenodd" d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z" clip-rule="evenodd" />
+            </svg>
```
Comment (dsandoval11) [inline components/books-search.js]: download the svg image and use `img` tag

Comment (jsequea) [reply components/books-search.js]: All changes succesfull

**Review by dsandoval11 — APPROVED** [review]

**Pull request merged**